blog/content/post/2020-08-05-PCIe-features.md

```diff
@@ -0,0 +1,177 @@
+
+**Baseline Error Reporting**
+
+Bits corresponding to this capability are located partly in the configuration
+header, partly in the Capability register block. All of them must be present in
+all PCIe functions.
+
+- **Power Management**
+
+It permits the OS to manage the power environment of a Function directly by
+accessing registers corresponding to this capability. OS can set device stat to
+one of 4 states. 2 of them are mandatory, and 2 can be optionally implemented.
+D0 state is a state where the function is fully operational and uses full power,
+every next state (D1,D2,D3) uses lower power, but also takes more time to recover
+to D0.
+
+- **Message Signaled Interrupts**
+
+PCI has pins to let the central interrupt controller know that it needs service.
+It significantly increased the number of needed lines. In PCIe instead of
+additional lines special messages are used to signal interrupt.
+These “messages” are actually memory writes to addresses corresponding to a given
+function. Address of memory to write and data being written are located in MSI
+capability structure.
+
+___
+
+**All following capabilities are optional.**
+
+___
+
+- **EXtended Message Signaled Interrupts**
+
+Extends MSI from 32 possible vectors to 2048 by placing a table containing
+addresses and messages of each vector in RAM instead of device registers.
+
+- **DYNAMIC POWER ALLOCATION**
+
+This capability extends power states by additional 32 states between D0 and D1.
+
+- **Power Budgeting**
+
+Its goal is to allocate power for PCIe hot plug devices that are added to the
+system during runtime. This ensures that the system can allocate the proper
+amount of power and cooling for these devices.
+
+- **Advanced Error Reporting**
+
+It provides additional registers that give error handling software more
+information to work with in diagnosing and recovering from problems.
+
+- **Virtual Channel**
+
+It adds buffers that act as queues for outgoing packets. This capability allows
+Device to assign one of 8 priorities to packets. Every priority has its own
+buffer. Packets from buffers with higher priorities are sent more frequently.
+
+- **TLP Processing Hints**
+
+It adds the possibility to use cache inside switches and the root complex to
+alleviate the need for RAM usage.
+
+- **Resizable BAR Capability**
+
+It replaces the BAR registers from HEADER with bigger registers in this
+capability structure, allowing bigger space allocation for functions.
+
+- **Access Control Services**
+
+The PCIe specification allows for peer-to-peer transactions.  This means that it
+is possible and even desirable in some cases for one PCIe endpoint to send data
+directly to another endpoint without having to go through the Root Complex.
+ACS provides a mechanism by which a Peer-to-Peer PCIe transaction can be forced
+to go up through the PCIe Root Complex. ACS can be thought of as a kind of
+gate-keeper - preventing unauthorized transactions from occurring.
```

Comment:
This is explained very well. Try to do something similar in other capabilities as well